src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,174 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(Node<T> prev, T element, Node<T> next) {
+            this.prev = prev;
+            this.item = element;
+            this.next = next;
+        }
+    }
+
+    private void checkIndex(int index, int size) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("Index is not valid. Index: "
+                    + index + ". Size of list: " + size);
+        }
+    }
+
+    private void addMiddle(T value, int index) {
+        Node middleNode = new Node(findNode(index - 1), value, findNode(index));
+        findNode(index - 1).next = middleNode;
+        findNode(index).prev = middleNode;
```

Comment:
4 iterations over the list
if you have some node - you can access the previous node with node.prev